### PR TITLE
Fix StaleElementReferenceException

### PR DESCRIPTION
**Pull request by jiuguangw**
Fix the StaleElementReferenceException error mentioned in #2768 #2780 #2772 (maybe others as well). Created based on comment by @uluQulu in #2745. Worked for me as of August 27.

### CODE REVIEW

**Comment by CharlesCCC** [comment]
chang looks good to me, @jiuguangw thank you for the quick PR and contribution.

File: instapy/unfollow_util.py [hunk 1]
```diff
@@ -723,9 +723,19 @@ def get_users_through_dialog(browser,
                                                          logger,
                                                          logfolder)
                     simulated_list.extend(quick_follow)
+                    # declare the dialog box once again after the DOM change
+                    dialog_address = "//div[text()='Followers' or text()='Following']/../../following-sibling::div"
```
Comment (uluQulu) [inline instapy/unfollow_util.py]: `dialog_adress` is already declared above 

-----------

Cheers 😁
---------

**Comment by samihamine** [comment]
I'v just tested, It works like a charm 👍 Thanks @jiuguangw

**Comment by tompicca** [comment]
Just tested, works great!

**Comment by tompicca** [comment]
I notice that this update takes 5 minutes each time a user no longer exist when unfollowing:
```
ERROR [2018-08-28 12:30:31] [empire.pumpshoes]  --> Unfollow error with b'_k4ti4_', maybe username has changed or he/she blocked you...
ERROR [2018-08-28 12:30:31] [empire.pumpshoes]  --> Unfollow error with b'_k4ti4_', maybe no longer exists...
ERROR [2018-08-28 12:35:33] [empire.pumpshoes]  --> Unfollow error with b'chrisss_ttinaa', maybe username has changed or he/she blocked you...
ERROR [2018-08-28 12:35:33] [empire.pumpshoes]  --> Unfollow error with b'chrisss_ttinaa', maybe no longer exists...
ERROR [2018-08-28 12:40:35] [empire.pumpshoes]  --> Unfollow error with b'tomasone_antonella', maybe username has changed or he/she blocked you...
ERROR [2018-08-28 12:40:35] [empire.pumpshoes]  --> Unfollow error with b'tomasone_antonella', maybe no longer exists...
```
There's a way to fix that?

**Comment by CharlesCCC** [comment]
@tompicca open your followedpool.csv and remove any users that have a followed date larger than your setting. (those are the users who might have changed their username or blocked you). so the Instapy can't go to their profile page and unfollow them.

**Comment by tompicca** [comment]
@CharlesCCC,
thanks for your reply, unfortunately I think the problem is elsewhere. I'm receiving this error with almost every profile I'm trying to unfollow. I checked by myself and despite the error, these profiles are normally working and never blocked me. 🤔
```
ERROR [2018-08-28 16:17:01] [empire.pumpshoes]  --> Unfollow error with b'a.n.d.r.e.e.a.h._', maybe username has changed or he/she blocked you...
ERROR [2018-08-28 16:17:01] [empire.pumpshoes]  --> Unfollow error with b'a.n.d.r.e.e.a.h._', maybe no longer exists...
ERROR [2018-08-28 16:22:02] [empire.pumpshoes]  --> Unfollow error with b'vash_tanec', maybe username has changed or he/she blocked you...
ERROR [2018-08-28 16:22:02] [empire.pumpshoes]  --> Unfollow error with b'vash_tanec', maybe no longer exists...
ERROR [2018-08-28 16:27:05] [empire.pumpshoes]  --> Unfollow error with b'alexnesporova', maybe username has changed or he/she blocked you...
ERROR [2018-08-28 16:27:05] [empire.pumpshoes]  --> Unfollow error with b'alexnesporova', maybe no longer exists..
```

**Comment by CharlesCCC** [comment]
Can you please share your followedpool with us ? 

another thing you can do is, can you monitor the Instapy running ? because I want to know if you are seeming the profile page for each unfollowed user or not.

**Comment by tompicca** [comment]
Of course
[empire.pumpshoes_followedPool.txt](https://github.com/timgrossmann/InstaPy/files/2328563/empire.pumpshoes_followedPool.txt)

**Comment by CharlesCCC** [comment]
what was your unfollowed setting ? how long was the delay before unfollow ?

**Comment by tompicca** [comment]
Here you are:
`session.unfollow_users(amount=1000, allFollowing=True, unfollow_after=1*60*60, sleep_delay=200)`

**Comment by CharlesCCC** [comment]
so, 
by look at your file and your setting. 

you have `unfollow_after=1*60*60` which means it will unfollow any user after 1 hour after you followed them. 

but look at the file you provided, all the user are from date 8-23 to 8-24, so it means, those should have been unfollow long before the change. which means, those user has been there for a long time now. 

and by look at this user `alexnesporova` , the user profile do exist, so it likely your got soft-banned from instagram by unfollowing too many account. I believe there is a limit amount of user you can unfollow each hour/day.  please look that up. 

one more things. unfollow amount 1000 seems quite a lots, you might want reduce that, and that probably is the reason you got banned.

**Comment by Cippoco** [comment]
Hi guys, i've downloaded the fix and edit the follow_util but form me the fix doesn't work, attached the log:

> Traceback (most recent call last):
>  File "quickstart.py", line 13, in <module>
>  session.follow_user_followers(['...'], amount=5, sleep_delay=36, randomize=True, interact=True)
>  File "C:\Users\Administrator\Desktop\InstaPy-master\instapy\instapy.py", line 2052, in follow_user_followers
>  self.logfolder)
>  File "C:\Users\Administrator\Desktop\InstaPy-master\instapy\unfollow_util.py", line 924, in get_given_user_followers
>  channel, logger, logfolder)
>  File "C:\Users\Administrator\Desktop\InstaPy-master\instapy\unfollow_util.py", line 712, in get_users_through_dialog
>  quick_username = dialog_username_extractor(quick_button)
>  File "C:\Users\Administrator\Desktop\InstaPy-master\instapy\unfollow_util.py", line 760, in dialog_username_extractor
>  if person and hasattr(person, 'text') and person.text:
>  File "C:\Program Files (x86)\Python\Python36-32\lib\site-packages\selenium\webdriver\remote\webelement.py", line 76, in text
>  return self._execute(Command.GET_ELEMENT_TEXT)['value']
>  File "C:\Program Files (x86)\Python\Python36-32\lib\site-packages\selenium\webdriver\remote\webelement.py", line 628, in _execute
>  return self._parent.execute(command, params)
>  File "C:\Program Files (x86)\Python\Python36-32\lib\site-packages\selenium\webdriver\remote\webdriver.py", line 312, in execute
>  self.error_handler.check_response(response)
>  File "C:\Program Files (x86)\Python\Python36-32\lib\site-packages\selenium\webdriver\remote\errorhandler.py", line 242, in check_response
>  raise exception_class(message, screen, stacktrace)
>  selenium.common.exceptions.StaleElementReferenceException: Message: stale element reference: element is not attached to the page document
>  (Session info: chrome=68.0.3440.106)
>  (Driver info: chromedriver=2.40.565498 (ea082db3280dd6843ebfb08a625e3eb905c4f5ab),platform=Windows NT 10.0.14393 x86_64)

**Comment by uluQulu** [comment]
@cippoco 
Can you attach your currently used problematic **unfollow_util.py** file in a _ZIP_?

**Comment by Cippoco** [comment]
@uluQulu 
Here, rename it! Tnx

[unfollow_util.txt](https://github.com/timgrossmann/InstaPy/files/2328717/unfollow_util.txt)

**Comment by uluQulu** [comment]
@cippoco 
It seems you have not applied https://github.com/timgrossmann/InstaPy/issues/2774#issuecomment-416583797.

Apply it. It will solve all of your problems.

-------

Cheers 😁
-----------

**Comment by tompicca** [comment]
@CharlesCCC,
I supposed that `unfollow_after=1*60*60` means "do not unfollow before 1 hour from follow".
 
But you said "it will unfollow any user after 1 hour after you followed them." and " those user has been there for a long time now." I'm a bit confused, Is there a range of time I can unfollow users? I never got problems if I unfollowed users 1 or 2 weeks after follow them.

Also, I can assure you I'm not banned, I just manually unfollow some users included `alexnesporova` without any problem.

**Comment by CharlesCCC** [comment]
@tompicca  it means "unfollow user after it has been followed more than 3600 seconds" 

what I meant by ` " those user has been there for a long time now."` is to saying, those user has been in your file for couple days now (far more than 3600 second limit you set). so it should has been all unfollowed if you didn't get "soft-banned" 

and yes, you got `soft-banned`. Have seem so many of those case, you can manually do it on the phone, but not through the same ip with website.

**Comment by tompicca** [comment]
@CharlesCCC 
Unfortunately, I can't agree with you my friend. The issue appears with a lot of users, but not definitely with all of them:
```
ERROR [2018-08-28 17:44:46] [empire.pumpshoes]  --> Unfollow error with b'alessandra_gargano_', maybe no longer exists...
ERROR [2018-08-28 17:45:14] [empire.pumpshoes]  --> Unfollow error with b'lyaa.lh', maybe no longer exists....
ERROR [2018-08-28 17:45:14] [empire.pumpshoes]  --> Unfollow error with b'lyaa.lh', maybe username has changed or he/she blocked you...
ERROR [2018-08-28 17:45:14] [empire.pumpshoes]  --> Unfollow error with b'lyaa.lh', maybe no longer exists...
INFO [2018-08-28 17:45:23] [empire.pumpshoes]  --> Ongoing Unfollow From InstaPy 1/582, now unfollowing: b'panettam80'
INFO [2018-08-28 17:45:23] [empire.pumpshoes]   Removed '2018-08-23 07:11 ~ panettam80' from followedPool.csv file

INFO [2018-08-28 17:45:45] [empire.pumpshoes]  --> Ongoing Unfollow From InstaPy 2/582, now unfollowing: b'_k4ti4_'
INFO [2018-08-28 17:45:45] [empire.pumpshoes]   Removed '2018-08-23 07:20 ~ _k4ti4_' from followedPool.csv file

INFO [2018-08-28 17:46:11] [empire.pumpshoes]  --> Ongoing Unfollow From InstaPy 3/582, now unfollowing: b'chrisss_ttinaa'
INFO [2018-08-28 17:46:11] [empire.pumpshoes]   Removed '2018-08-23 07:21 ~ chrisss_ttinaa' from followedPool.csv file

INFO [2018-08-28 17:46:41] [empire.pumpshoes]  --> Ongoing Unfollow From InstaPy 4/582, now unfollowing: b'tomasone_antonella'
INFO [2018-08-28 17:46:41] [empire.pumpshoes]   Removed '2018-08-23 07:21 ~ tomasone_antonella' from followedPool.csv file

INFO [2018-08-28 17:47:07] [empire.pumpshoes]  --> Ongoing Unfollow From InstaPy 5/582, now unfollowing: b'federica.fasolo'
INFO [2018-08-28 17:47:07] [empire.pumpshoes]   Removed '2018-08-23 07:22 ~ federica.fasolo' from followedPool.csv file
```
This means that I'm not soft-banned I suppose. 
Also I'm running InstaPy on my computer's terminal, and I manually unfollowed users with chrome for desktop, which I suppose have the same ip.
Let me say that the problem is definitely elsewhere.

**Comment by CharlesCCC** [comment]
@tompicca let's found out where the problem is, do me a favor. keep  eyes on the browser when it can't unfollow someone,  share here on what you see.

**Comment by Cippoco** [comment]
@uluQulu 
Yes, now i've solve that error but another one is returned:

selenium.common.exceptions.NoSuchElementException: Message: no such element: Unable to locate element: {"method":"xpath","selector":"//div[text()='Followers' or text()='Following']/../../following-sibling::div"}
 (Session info: chrome=68.0.3440.106)
 (Driver info: chromedriver=2.40.565498 (ea082db3280dd6843ebfb08a625e3eb905c4f5ab),platform=Windows NT 10.0.14393 x86_64)

Reading on the other topics appear to depend on the instapy repo version, but i've downloaded the last one...

**Comment by uluQulu** [comment]
@cippoco 
Please share the full traceback also attach your **unfollow_util.py**

**Comment by Cippoco** [comment]
@uluQulu 

Here the unfollow_util and full log

[unfollow_util.txt](https://github.com/timgrossmann/InstaPy/files/2329735/unfollow_util.txt)

> File "quickstart.py", line 13, in <module>
>  session.follow_user_followers(['...l'], amount=5, sleep_delay=36, randomize=True, interact=True)
>  File "C:\Users\Administrator\Desktop\InstaPy-master\instapy\instapy.py", line 2052, in follow_user_followers
>  self.logfolder)
>  File "C:\Users\Administrator\Desktop\InstaPy-master\instapy\unfollow_util.py", line 919, in get_given_user_followers
>  channel, logger, logfolder)
>  File "C:\Users\Administrator\Desktop\InstaPy-master\instapy\unfollow_util.py", line 729, in get_users_through_dialog
>  dialog = browser.find_element_by_xpath(dialog_address)
>  File "C:\Program Files (x86)\Python\Python36-32\lib\site-packages\selenium\webdriver\remote\webdriver.py", line 385, in find_element_by_xpath
>  return self.find_element(by=By.XPATH, value=xpath)
>  File "C:\Program Files (x86)\Python\Python36-32\lib\site-packages\selenium\webdriver\remote\webdriver.py", line 955, in find_element
>  'value': value})['value']
>  File "C:\Program Files (x86)\Python\Python36-32\lib\site-packages\selenium\webdriver\remote\webdriver.py", line 312, in execute
>  self.error_handler.check_response(response)
>  File "C:\Program Files (x86)\Python\Python36-32\lib\site-packages\selenium\webdriver\remote\errorhandler.py", line 242, in check_response
>  raise exception_class(message, screen, stacktrace)
>  selenium.common.exceptions.NoSuchElementException: Message: no such element: Unable to locate element: {"method":"xpath","selector":"//div[text()='Followers' or text()='Following']/../../following-sibling::div"}
>  (Session info: chrome=68.0.3440.106)
>  (Driver info: chromedriver=2.40.565498 (ea082db3280dd6843ebfb08a625e3eb905c4f5ab),platform=Windows NT 10.0.14393 x86_64)
>

**Comment by uluQulu** [comment]
@cippoco
Does it happens always?
Can you watch the browser and tell me if it opens a new tab and then returns back properly and where it crashes?

**Comment by Cippoco** [comment]
@uluQulu 
Yes, it happens always, launching the bot in "follow user follower" mode, it begin work, it open chrome and logging in, liking some picture, follows 2 people random, when it switch on the first user on the profile list given, it crashes when try to follow the firsth profile...

**Comment by uluQulu** [comment]
@cippoco 
liking some picture? 
I think you are having another issue than I thought. Cos while getting usernames for `follow_user_followers` it cannot like 😂
Can you attach the set of actions from your **quickstart** script? (_remove personal information_)

**Comment by Cippoco** [comment]
I'm using the gui, but the only check selected is the "follow_user_followers", other optiona are all blanck, but as you seen in the quickstart reamain, attached:

```
from instapy import InstaPy

session = InstaPy(username='...', password='...')

session.login()

session.set_ignore_users(['...'])
session.set_do_follow(enabled=True, percentage=30, times=4)
session.set_user_interact(amount=15, randomize=True, percentage=20)

session.follow_user_followers(['... '], amount=5, sleep_delay=36, randomize=True, interact=True)

session.end()
```

**Comment by uluQulu** [comment]
@cippoco 
Unfortunately I cannot see what is going on in your browser window.
Try to help me understand the problem by answering these questions 🙋🏼‍♂️
+ Does it open dialog box properly? ["_Followers_" dialog box of the first user in `follow_user_followers()`]
+ Does it scroll down properly?
+ Does it do simulated follow? [clicking on a random follow button- user `A`]
+ Does it open a new tab?
+ Does it navigate to the profile page of user `A`? [in the newly opened tab]
+ Does it reload the page successfully? [and wait then a second to get user ID]
+ Does it close the newly opened tab successfully?
+ Does it continue to scroll down properly?

It must crash in one of the steps above. Which one it is? [accurate answer please]

------

Cheers 🤓
--------

**Comment by uluQulu** [comment]
@cippoco 
Although I am unsure about your problem, you can use the newly written fix at https://github.com/timgrossmann/InstaPy/issues/2774#issuecomment-416583797
> **5**th & **6**th steps are new, apply them.

---------

Cheers 😁
----------

**Comment by SammyHam** [comment]
> @CharlesCCC
> Unfortunately, I can't agree with you my friend. The issue appears with a lot of users, but not definitely with all of them:
> 
> ERROR [2018-08-28 17:44:46] [empire.pumpshoes]  --> Unfollow error with b'alessandra_gargano_', maybe no longer exists...
> ERROR [2018-08-28 17:45:14] [empire.pumpshoes]  --> Unfollow error with b'lyaa.lh', maybe no longer exists....
> ERROR [2018-08-28 17:45:14] [empire.pumpshoes]  --> Unfollow error with b'lyaa.lh', maybe username has changed or he/she blocked you...
> ERROR [2018-08-28 17:45:14] [empire.pumpshoes]  --> Unfollow error with b'lyaa.lh', maybe no longer exists...
> INFO [2018-08-28 17:45:23] [empire.pumpshoes]  --> Ongoing Unfollow From InstaPy 1/582, now unfollowing: b'panettam80'
> INFO [2018-08-28 17:45:23] [empire.pumpshoes]   Removed '2018-08-23 07:11 ~ panettam80' from followedPool.csv file
> 
> INFO [2018-08-28 17:45:45] [empire.pumpshoes]  --> Ongoing Unfollow From InstaPy 2/582, now unfollowing: b'_k4ti4_'
> INFO [2018-08-28 17:45:45] [empire.pumpshoes]   Removed '2018-08-23 07:20 ~ _k4ti4_' from followedPool.csv file
> 
> INFO [2018-08-28 17:46:11] [empire.pumpshoes]  --> Ongoing Unfollow From InstaPy 3/582, now unfollowing: b'chrisss_ttinaa'
> INFO [2018-08-28 17:46:11] [empire.pumpshoes]   Removed '2018-08-23 07:21 ~ chrisss_ttinaa' from followedPool.csv file
> 
> INFO [2018-08-28 17:46:41] [empire.pumpshoes]  --> Ongoing Unfollow From InstaPy 4/582, now unfollowing: b'tomasone_antonella'
> INFO [2018-08-28 17:46:41] [empire.pumpshoes]   Removed '2018-08-23 07:21 ~ tomasone_antonella' from followedPool.csv file
> 
> INFO [2018-08-28 17:47:07] [empire.pumpshoes]  --> Ongoing Unfollow From InstaPy 5/582, now unfollowing: b'federica.fasolo'
> INFO [2018-08-28 17:47:07] [empire.pumpshoes]   Removed '2018-08-23 07:22 ~ federica.fasolo' from followedPool.csv file
> This means that I'm not soft-banned I suppose.
> Also I'm running InstaPy on my computer's terminal, and I manually unfollowed users with chrome for desktop, which I suppose have the same ip.
> Let me say that the problem is definitely elsewhere.

**Comment by samihamine** [comment]
> @CharlesCCC
> Unfortunately, I can't agree with you my friend. The issue appears with a lot of users, but not definitely with all of them:
> 
```
> ERROR [2018-08-28 17:44:46] [empire.pumpshoes]  --> Unfollow error with b'alessandra_gargano_', maybe no longer exists...
> ERROR [2018-08-28 17:45:14] [empire.pumpshoes]  --> Unfollow error with b'lyaa.lh', maybe no longer exists....
> ERROR [2018-08-28 17:45:14] [empire.pumpshoes]  --> Unfollow error with b'lyaa.lh', maybe username has changed or he/she blocked you...
> ERROR [2018-08-28 17:45:14] [empire.pumpshoes]  --> Unfollow error with b'lyaa.lh', maybe no longer exists...
> INFO [2018-08-28 17:45:23] [empire.pumpshoes]  --> Ongoing Unfollow From InstaPy 1/582, now unfollowing: b'panettam80'
> INFO [2018-08-28 17:45:23] [empire.pumpshoes]   Removed '2018-08-23 07:11 ~ panettam80' from followedPool.csv file
> 
> INFO [2018-08-28 17:45:45] [empire.pumpshoes]  --> Ongoing Unfollow From InstaPy 2/582, now unfollowing: b'_k4ti4_'
> INFO [2018-08-28 17:45:45] [empire.pumpshoes]   Removed '2018-08-23 07:20 ~ _k4ti4_' from followedPool.csv file
> 
> INFO [2018-08-28 17:46:11] [empire.pumpshoes]  --> Ongoing Unfollow From InstaPy 3/582, now unfollowing: b'chrisss_ttinaa'
> INFO [2018-08-28 17:46:11] [empire.pumpshoes]   Removed '2018-08-23 07:21 ~ chrisss_ttinaa' from followedPool.csv file
> 
> INFO [2018-08-28 17:46:41] [empire.pumpshoes]  --> Ongoing Unfollow From InstaPy 4/582, now unfollowing: b'tomasone_antonella'
> INFO [2018-08-28 17:46:41] [empire.pumpshoes]   Removed '2018-08-23 07:21 ~ tomasone_antonella' from followedPool.csv file
> 
> INFO [2018-08-28 17:47:07] [empire.pumpshoes]  --> Ongoing Unfollow From InstaPy 5/582, now unfollowing: b'federica.fasolo'
> INFO [2018-08-28 17:47:07] [empire.pumpshoes]   Removed '2018-08-23 07:22 ~ federica.fasolo' from followedPool.csv file
```
> This means that I'm not soft-banned I suppose.
> Also I'm running InstaPy on my computer's terminal, and I manually unfollowed users with chrome for desktop, which I suppose have the same ip.
> Let me say that the problem is definitely elsewhere.

@CharlesCCC @tompicca Seems like be a problem with private accounts for which an answer to our request to follow is pending (status "Requested") BUT not all of them. This problem is present only when we are using the unfollowing option "InstapyFollowed=(True, "all")"

**Comment by samihamine** [comment]
@CharlesCCC  after a little debug, seems like that for some accounts the function "get_following_status" generates an exception and the loop calls the first account having an userid in the file "xxx_followedPool.csv" (following your last evolution) then returns to the following account in the list... 
Any ideas ?

**Comment by uluQulu** [comment]
@SammyHam 
I solved a few misbehaviours in the unfollow mechamism at my latest PR.
I recommend you to download the fork with which I opened the PR and try again.
If it persists, then there is a problem.

>You will only put your **quickstart** script into the downloaded fork, everything else is the same.
Also you should install plyer from pip :P

------

Cheers 😁
------

**Comment by samihamine** [comment]
@uluQulu  works like a charm ✌️ no problem found after about ten minutes! I let it run tonight ... Thanks for your help.
Your last PR is huuuge, a lot of work! You rock dude 👍

**Comment by uluQulu** [comment]
Thank you so much @SammyHam 😄
>Sometimes logs says it works fine but it does not in actual. I mean make sure to check in graphical mode, too, that it really does what it says..

If so, it means there is no serious error, @CharlesCCC.

--------

Cheers 😁
-------

**Comment by boldestfortune** [comment]
Hey, downloaded your unfollow.py scripts. works like a charm but I noticed some additional comments at start not sure if this is related?

**Pull request merged**